### PR TITLE
feat(incremental-planner): add SharedPath and SelectionBuilder

### PR DESCRIPTION
**Pull request by tninesling**
## Summary

Introduces two data structures used throughout the incremental planner:

- **`SharedPath<T>`**: A persistent cons-list giving O(1) clone and extend. Paths are forked thousands of times per BULB probe (sibling selections share a prefix) but materialized only once on the winning plan. Making it `pub` enables the doc test to run in CI without `ignore`.

- **`SelectionBuilder`**: An append-only log of (path, selection) entries accumulated during search. Insert is O(path_depth); clone is O(n). Entries are materialized into a `SelectionSet` only once, when the winning plan is finalized. Checkpoint/restore truncates the log for undo during backtracking.

Also stubs the `fetch_graph` module (populated in the next PR).

## Test plan

- `SharedPath` doc test exercises the sharing/forking invariant
- Unit tests cover `parent`, empty path, and single-element edge cases
- `SelectionBuilder` unit tests cover clone-snapshot isolation and `field_signatures` recursion

### CODE REVIEW

**Comment by dariuszkuc** [comment]
Nothing serious but might be worth it to address as well

------
### 1. `field_signatures` keys are not response paths, so the check it exists for is defeated
`fetch_graph/selection_builder.rs:96` and `:114` (doc contract at `:75-79`)

The doc says:

> Two builders assigning different signatures to the same response path cannot be materialized into
> one fetch because the same response key would need two different field executions.

But inline fragments contribute a key segment at both sites:

```rust
// :113-115, walking the entry's op path
OpPathElement::InlineFragment(frag) => {
    prefix = format!("{prefix}/{frag}");
}
// :95-98, walking a recorded selection set
Selection::InlineFragment(frag_sel) => {
    let key = format!("{prefix}/{}", frag_sel.inline_fragment);
    record_selection_set(out, &key, &frag_sel.selection_set);
}
```

A fragment contributes no level to the GraphQL response, so these keys are not response paths.
`node { name }` keys `/node/name`; `node { ... on User { name: fullName } }` keys
`/node/... on User/name`. Those are the same response key at the same response position with two
different field executions — the exact conflict the map is built to catch — and it is missed. Since
type-conditioned selections are the common case in a federated planner, the check is defeated
roughly whenever it matters.

Two things to note before fixing:

- **The prescription is not just "drop the fragment segment."** Per the field-merging rules, two
  same-key fields under *provably disjoint* object type conditions are legal with different
  arguments, so keying on the bare response path would over-report those as conflicts. The
  distinction that has to survive is unconditioned-vs-conditioned and overlapping-vs-disjoint type
  conditions, not "fragment or not."
- **A test in this PR asserts the current behavior**, so a fix has to update it rather than being
  caught by it: `field_signatures_recurses_into_sub_selections_and_fragments` (`:152`) asserts
  `address_key.contains("... on User")` at `:201`.

At minimum, if this is intended for now, the doc comment at `:75-79` should say what the keys
actually are and what class of conflict the map therefore does not detect.

**Why it matters downstream:** #10148 (`fetch_graph/mod.rs:728`) uses `field_signatures` as the
merge-compatibility predicate for sibling entity groups. With this keying, `value(scale: 100)` and
`... on Admin { value }` are judged compatible, merged, and produce a field-merging conflict in the
generated subgraph operation.

### 2. `field_signatures` is last-write-wins, so it also hides conflicts *within* one builder
`fetch_graph/selection_builder.rs:90` and `:110`

```rust
out.insert(key.clone(), field_sel.field.to_string());
```

Both insert sites overwrite silently. One builder accumulating `value` from one commit and
`value(scale: 100)` from another under the same path ends up reporting a single signature, and no
caller can see that two were recorded. Since `merge_from` (`:71`) concatenates logs, the merged
builder has the same blind spot.

Suggestion: `match out.entry(key)` and, on an occupied slot with a differing signature, surface the
conflict — either by returning `Result<HashMap<..>, _>` or by having the function return the
conflicting pairs alongside the map.

### 3. `merge_from` can violate the invariant it never checks
`fetch_graph/selection_builder.rs:71-73`

```rust
pub(super) fn merge_from(&mut self, other: &SelectionBuilder) {
    self.entries.extend(other.entries.iter().cloned());
}
```

`field_signatures` exists to decide whether two builders *may* be merged, and `merge_from`
unconditionally concatenates without consulting it. Nothing in this PR calls either, so there is no
call site to point at yet — which is why it is worth pinning the contract now, while it is free:
either return `Result` and check compatibility internally, or document "caller must have verified
compatibility via `field_signatures` first" on `merge_from`.

(For reference, when the first caller does arrive in #10148 it checks compatibility — via the
predicate from item 1.)

### 4. Nit — `SharedPath` drops recursively
`shared_path.rs:41-45`

```rust
struct Node<T> {
    value: T,
    next: Option<Arc<Node<T>>>,
}
```

The classic cons-list drop hazard: dropping the last owner of a spine of length *n* recurses *n*
frames, and there is no `Drop` impl. In practice depth here is operation nesting depth, which the
router bounds at `parser_max_recursion` (default 500), so this is hardening rather than a live bug —
but an iterative `Drop` that unlinks `next` in a loop is ~10 lines and removes the class outright,
and it is cheaper to add now than after the type is load-bearing.

While in the file: `iter()` (`:84-93`) allocates a `Vec` per call, so a caller that iterates
per-entry (e.g. `field_signatures`) pays an allocation per entry on a type whose selling point is
O(1) clone/push. Fine at finalization boundaries, worth a doc note so nobody calls it in the search
loop.

### 5. Nit — `pub mod shared_path` puts an internal structure in the published API
`incremental_planner/mod.rs:2-5`

```rust
#[allow(dead_code)]
pub(crate) mod fetch_graph;
#[allow(dead_code)]
pub mod shared_path;
```

`query_plan` and `query_plan::incremental_planner` are both `pub`, so this exports
`apollo_federation::query_plan::incremental_planner::shared_path::SharedPath` from the crate
permanently. The only reason it needs to be `pub` is the doc test at `shared_path.rs:20-34`, which
imports it by that path. `#[doc(hidden)] pub mod`, or moving the example into a `#[test]` fn, keeps
the example running without the API commitment.

Also on those two attributes:

- `#[allow(dead_code)]` on `pub mod shared_path` is a no-op — the lint does not fire on
  reachable-`pub` items in the first place.
- On `pub(crate) mod fetch_graph` it suppresses dead-code warnings for that entire subtree
  indefinitely, and the subtree grows to most of the planner over the next 25 PRs. If it is only
  needed until the entry point lands, an item-level `#[allow]` on the specific unreached items (or
  an `expect` with a note) keeps the signal.

### 6. Nit — one test name overstates what it asserts
`fetch_graph/selection_builder.rs:130-136`

`empty_builder_clone_is_cheap` asserts that both builders are empty; it observes nothing about cost.
`clone_of_empty_builder_is_empty` says what it checks. (Same pattern the chain review flags in
#10108/#10151 — worth not starting it here.)

**Comment by tninesling** [comment]
Thanks for the thorough review. All addressed in fcdcfdf:

**1. field_signatures keys vs response paths**: Fixed. Inline fragments no longer contribute a path segment. Keys are now bare response paths. As you noted, this is conservative: fields under disjoint type conditions that legally differ will appear to conflict, but that's the safe direction (false conflict keeps groups separate, missed conflict produces an invalid subgraph query). Added a doc comment explaining this tradeoff.

**2. last-write-wins hiding intra-builder conflicts**: Fixed. Added a `try_insert` helper that returns false on a conflicting signature at the same key. `field_signatures` now returns `Option<HashMap<...>>`, returning `None` on internal conflict. Added a test (`field_signatures_detects_intra_builder_conflict`) that confirms two different signatures at the same response path are caught.

**3. merge_from contract**: Added doc on `merge_from`: "Caller must verify compatibility via `field_signatures` first."

**4. SharedPath iterative Drop**: Added an iterative `Drop` impl that unlinks `next` in a loop via `Arc::try_unwrap`. Also added a doc note on `iter()` about its allocation cost.

**5. pub mod shared_path**: Keeping this `pub` intentionally so the doc test works without `#[doc(hidden)]` gymnastics. We're not committed to API stability on this path.

**6. Test rename**: `empty_builder_clone_is_cheap` renamed to `clone_of_empty_builder_is_empty`.

File: apollo-federation/src/query_plan/incremental_planner/fetch_graph/selection_builder.rs [hunk 1]
```diff
@@ -0,0 +1,367 @@
+use std::collections::HashMap;
+use std::sync::Arc;
+
+use super::super::shared_path::SharedPath;
+use crate::operation::Selection;
+use crate::operation::SelectionSet;
+use crate::query_graph::graph_path::operation::OpPathElement;
+
+/// Opaque undo checkpoint: the entries Vec length at a point in time.
+/// Restoring truncates all insertions made since, O(1).
```
Comment (sachindshinde) [inline apollo-federation/src/query_plan/incremental_planner/fetch_graph/selection_builder.rs]: Truncating a `Vec` isn't generally O(1), as it runs `ptr::drop_in_place()` to run the destructors of the truncated elements. And in this case at least, the `Arc`s in `SelectionEntry` implement `Drop`.

File: apollo-federation/src/query_plan/incremental_planner/shared_path.rs [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+}
+
+impl<T> Drop for SharedPath<T> {
+    /// Iterative drop to avoid stack overflow on long spines.
+    fn drop(&mut self) {
+        let mut current = self.head.take();
+        while let Some(arc) = current {
+            // If we're the sole owner, unwrap and take the next pointer;
+            // Otherwise another SharedPath shares the tail, so stop.
+            match Arc::try_unwrap(arc) {
```
Comment (sachindshinde) [inline apollo-federation/src/query_plan/incremental_planner/shared_path.rs]: This is a misuse of `Arc::try_unwrap()` and can still cause stack overflow; you want to use `Arc::into_inner()` instead here (the docs talk about this more).

Specifically, if there are multiple concurrent calls to `Arc::into_inner()` on all `Arc` clones, it's guaranteed that exactly one of them returns the inner value. This is not true of `Arc::try_unwrap()`, and its possible for multiple concurrent calls to return `None`. That means that if e.g. two `SharedPath`s are being dropped simultaneously, and they encounter an `Arc` shared between them at the same time, both calls to `Arc::try_unwrap()` may return `Err`. In that case, the `Arc`s returned in the `Err`s are dropped normally, and whichever one occurs later will trigger the normal cascading behavior with `Arc` dropping (leading to stack overflow).

File: apollo-federation/src/query_plan/incremental_planner/fetch_graph/selection_builder.rs (hunk 1 above)
Comment (sachindshinde) [inline apollo-federation/src/query_plan/incremental_planner/fetch_graph/selection_builder.rs]: Truncating a `Vec` isn't generally O(1), as it runs `ptr::drop_in_place()` to run the destructors of the truncated elements. And in this case at least, the `Arc`s in `SelectionEntry` implement `Drop`.

File: apollo-federation/src/query_plan/incremental_planner/shared_path.rs [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+    /// finalization boundaries; avoid calling per-entry in hot loops.
+    pub fn iter(&self) -> Iter<'_, T> {
+        let mut nodes = Vec::with_capacity(self.len);
+        let mut current = &self.head;
+        while let Some(node) = current {
+            nodes.push(node.as_ref());
+            current = &node.next;
+        }
+        nodes.reverse();
+        Iter { nodes, pos: 0 }
```
Comment (sachindshinde) [inline apollo-federation/src/query_plan/incremental_planner/shared_path.rs]: Seems like you could just return `nodes.into_iter().map(|node| &node.value)` here (changing the return to `impl ExactSizeIterator<Item = &T>`) and avoid having to implement the custom `Iter`/iterator traits below.

File: apollo-federation/src/query_plan/incremental_planner/shared_path.rs [hunk 4]
```diff
@@ -0,0 +1,205 @@
...
+impl<T: Clone> SharedPath<T> {
+    /// Materialize into a Vec. O(n). Used at finalization boundaries.
+    pub fn to_vec(&self) -> Vec<T> {
+        self.iter().cloned().collect()
+    }
+
+    /// Build from a Vec.
+    pub fn from_vec(v: Vec<T>) -> Self {
+        let mut path = Self::new();
+        for item in v {
```
Comment (sachindshinde) [inline apollo-federation/src/query_plan/incremental_planner/shared_path.rs:116]: This consumes the vector `v`, so why would this need to be in `impl<T: Clone>` instead of the regular `impl`? (Doesn't seem like you use `T: Clone` here.)

File: apollo-federation/src/query_plan/incremental_planner/shared_path.rs [hunk 5]
```diff
@@ -0,0 +1,205 @@
...
+            path = path.pushed(item);
+        }
+        path
+    }
+}
+
+impl<T> Clone for SharedPath<T> {
```
Comment (sachindshinde) [inline apollo-federation/src/query_plan/incremental_planner/shared_path.rs]: This could be a `#[derive(Clone)]` on `SharedPath`.

File: apollo-federation/src/query_plan/incremental_planner/shared_path.rs [hunk 6]
```diff
@@ -0,0 +1,205 @@
...
+    fn clone(&self) -> Self {
+        Self {
+            head: self.head.clone(),
+            len: self.len,
+        }
+    }
+}
+
+impl<T> Default for SharedPath<T> {
```
Comment (sachindshinde) [inline apollo-federation/src/query_plan/incremental_planner/shared_path.rs:131]: This could be a `#[derive(Default)]` on `SharedPath` (and `Self::new()` could just return `Default::default()`).

File: apollo-federation/src/query_plan/incremental_planner/fetch_graph/selection_builder.rs [hunk 7]
```diff
@@ -0,0 +1,433 @@
...
+        debug_assert!(
+            cp.0 <= self.entries.len(),
+            "checkpoint is newer than the builder state; checkpoints must be restored in LIFO order",
+        );
+        self.entries.truncate(cp.0);
+    }
+
+    /// Absorb all entries from `other`, for post-search merging of sibling
+    /// entity groups. Caller must verify compatibility via
+    /// [`field_signatures`](Self::field_signatures) first.
```
Comment (sachindshinde) [inline apollo-federation/src/query_plan/incremental_planner/fetch_graph/selection_builder.rs]: Seeing
```rust
/// Caller must verify compatibility via [`field_signatures`](Self::field_signatures) first.
```
it seems this would apply to callers of `insert()` as well (probably worth a comment on `insert()`).

File: apollo-federation/src/query_plan/incremental_planner/fetch_graph/selection_builder.rs [hunk 8]
```diff
@@ -0,0 +1,433 @@
...
+    pub(super) fn merge_from(&mut self, other: &SelectionBuilder) {
+        self.entries.extend(other.entries.iter().cloned());
+    }
+
+    /// Map from response path to field signature (name, alias, arguments,
+    /// directives, and declared type). Two builders with different
+    /// signatures at the same response path cannot be merged into one fetch.
+    ///
+    /// The declared type is part of the signature because inline fragments
+    /// are transparent in the keys: same-named fields under different type
```
Comment (sachindshinde) [inline apollo-federation/src/query_plan/incremental_planner/fetch_graph/selection_builder.rs]:
```rust
/// The declared type is part of the signature because inline fragments
/// are transparent in the keys: same-named fields under different type
/// conditions land on the same key, and merging them is only valid when
/// their types agree (SameResponseShape).
```
That's not what the [spec validation rule](https://spec.graphql.org/draft/#SameResponseShape()) says. It's not about their types agreeing, it's about the response shapes of the selection sets (the types affect possible response shapes for two fields, but the types don't have to be the same).

File: apollo-federation/src/query_plan/incremental_planner/fetch_graph/selection_builder.rs [hunk 9]
```diff
@@ -0,0 +1,433 @@
...
+    /// conditions land on the same key, and merging them is only valid when
+    /// their types agree (SameResponseShape).
+    ///
+    /// Keys use field response names only; inline fragments are
+    /// transparent (they add no level to the response). Fields under
+    /// disjoint type conditions that legally differ per the field-merging
+    /// rules will appear to conflict here. That's conservative: a false
```
Comment (sachindshinde) [inline apollo-federation/src/query_plan/incremental_planner/fetch_graph/selection_builder.rs]:
```
/// That's conservative: a false
/// conflict keeps groups separate, while a missed conflict produces
/// an invalid subgraph query.
```
To make sure I understand the intended purpose of this function:
1. It's used to prevent potential conflicts in selection builders.
2. If there is a conflict, it definitely needs to report it.
3. It could mistakenly report a conflict when there isn't one.
4. We think that's okay because keeping groups separate should just mean a less efficient plan.

I can tell you the last statement isn't really true, because e.g. mutation fields can only be executed once (it matters if they end up in the same group or not), so presumably it can mean unsatisfiability at runtime (but maybe a later PR handles this specially).

File: apollo-federation/src/query_plan/incremental_planner/fetch_graph/selection_builder.rs [hunk 10]
```diff
@@ -0,0 +1,433 @@
...
+    /// an invalid subgraph query.
+    ///
+    /// Returns `None` if any response path has more than one distinct
+    /// field signature within this builder, which indicates a bug in the
+    /// commit logic.
+    pub(super) fn field_signatures(&self) -> Option<HashMap<String, String>> {
+        fn signature(field: &crate::operation::Field) -> String {
+            match field.field_position.get(field.schema.schema()) {
+                Ok(def) => format!("{field}: {}", def.ty),
+                Err(_) => field.to_string(),
```
Comment (sachindshinde) [inline apollo-federation/src/query_plan/incremental_planner/fetch_graph/selection_builder.rs]: Note that while the types don't need to agree, `SameResponseShape` does have requirements based on certain properties of field return types. Specifically, the non-null/list nesting has to match between fields, and if the return types are enum/scalar, they also have to match in type name. This means you can't really make guarantees about whether there's a conflict when you don't have type information, so I'd rather the function just return `None` in the `Err(_)` path here.

**Review by sachindshinde — CHANGES_REQUESTED** [review]

File: apollo-federation/src/query_plan/incremental_planner/fetch_graph/selection_builder.rs [hunk 11]
```diff
@@ -0,0 +1,433 @@
...
+            }
+        }
+
+        let mut out = HashMap::new();
+        for entry in &self.entries {
+            let mut prefix = String::new();
+            for element in entry.path.iter() {
+                match element.as_ref() {
+                    OpPathElement::Field(field) => {
+                        let key = format!("{prefix}/{}", field.response_name());
```
Comment (sachindshinde) [inline apollo-federation/src/query_plan/incremental_planner/fetch_graph/selection_builder.rs]: The stringified field signatures can have `/` as a character due to client-provided argument values. AFAICT it shouldn't result in distinct paths serializing to the same `key`, since the `/`s should be enclosed in quotes, but if downstream code tried to split based on `/` it would cause bugs. It would also cause bugs if e.g. the GraphQL spec decided to support field/argument descriptions one day (not to be confused with field definition/argument definition descriptions), which may be a thing in the future considering the spec recently updated to support descriptions on operations.

Is there something downstream that needs these to be stringified, or could we e.g. use a `Vec` of strings here?

File: apollo-federation/src/query_plan/incremental_planner/shared_path.rs (hunk 2 above)
Comment (tninesling) [reply apollo-federation/src/query_plan/incremental_planner/shared_path.rs]: Yeah, this was exactly the race. Fixed with Arc::into_inner in 680d811f8, with a spin-aligned two-thread drop test that SIGABRTed on the old code.

File: apollo-federation/src/query_plan/incremental_planner/fetch_graph/selection_builder.rs (hunk 1 above)
Comment (tninesling) [reply apollo-federation/src/query_plan/incremental_planner/fetch_graph/selection_builder.rs]: Right, the truncate drops each removed entry. Removed the O(1) claim in c888dd114.

File: apollo-federation/src/query_plan/incremental_planner/shared_path.rs (hunk 3 above)
Comment (tninesling) [reply apollo-federation/src/query_plan/incremental_planner/shared_path.rs]: Done in c888dd114. iter now returns impl ExactSizeIterator<Item = &T> and the custom iterator is gone.

File: apollo-federation/src/query_plan/incremental_planner/shared_path.rs (hunk 4 above)
Comment (tninesling) [reply apollo-federation/src/query_plan/incremental_planner/shared_path.rs:116]: Done, moved to the unbounded impl in c888dd114.

File: apollo-federation/src/query_plan/incremental_planner/shared_path.rs (hunk 5 above)
Comment (tninesling) [reply apollo-federation/src/query_plan/incremental_planner/shared_path.rs]: Done in c888dd114.

File: apollo-federation/src/query_plan/incremental_planner/shared_path.rs (hunk 6 above)
Comment (tninesling) [reply apollo-federation/src/query_plan/incremental_planner/shared_path.rs:131]: Deriving bounds T: Default, and we default paths of non-Default elements like SharedPath<Arc<OpPathElement>> in later PRs, so the manual impl stays. Added a test with a non-Default element type to pin that in c888dd114.

File: apollo-federation/src/query_plan/incremental_planner/fetch_graph/selection_builder.rs (hunk 7 above)
Comment (tninesling) [reply apollo-federation/src/query_plan/incremental_planner/fetch_graph/selection_builder.rs]: Moot now: field_signatures had no consumer in this stack, so c888dd114 removes it from this PR. It will be introduced with the sibling-merge pass that uses it.

File: apollo-federation/src/query_plan/incremental_planner/fetch_graph/selection_builder.rs (hunk 8 above)
Comment (tninesling) [reply apollo-federation/src/query_plan/incremental_planner/fetch_graph/selection_builder.rs]: You're right, the spec compares response shapes, not type names. This function moved out of the PR entirely (see the thread below), and the reworked version carries corrected wording.

File: apollo-federation/src/query_plan/incremental_planner/fetch_graph/selection_builder.rs (hunk 9 above)
Comment (tninesling) [reply apollo-federation/src/query_plan/incremental_planner/fetch_graph/selection_builder.rs]: Since there's no consumer until a much later PR, I'm moving the field_signatures bit to that PR where we can see how the invariants are held in the consumer. Feedback noted so I can fix up the implementation on that later branch.

File: apollo-federation/src/query_plan/incremental_planner/fetch_graph/selection_builder.rs (hunk 10 above)
Comment (tninesling) [reply apollo-federation/src/query_plan/incremental_planner/fetch_graph/selection_builder.rs]: Agreed the fallback was too loose. The reworked version (moving to the sibling-merge PR with the rest of field_signatures) propagates a resolution failure as an internal error and keeps None strictly for conflicts.

File: apollo-federation/src/query_plan/incremental_planner/shared_path.rs (hunk 5 above)
Comment (dariuszkuc) [reply apollo-federation/src/query_plan/incremental_planner/shared_path.rs]: fyi: Claude flags this as problematic and cites the same argument as for keeping custom `impl Default` below

File: apollo-federation/src/query_plan/incremental_planner/shared_path.rs (hunk 6 above)
Comment (sachindshinde) [reply apollo-federation/src/query_plan/incremental_planner/shared_path.rs:131]: In that case, you could use `#[derive_where(Default)]` to avoid the trait bound on `T`, via the [derive-where crate](https://crates.io/crates/derive-where). (Though this is relatively minor.)

**Review by sachindshinde — APPROVED** [review]
The rest are minor things, other than another comment on `field_signatures()`, but (as noted above) that code's been moved out of this PR into a later one.

File: apollo-federation/src/query_plan/incremental_planner/shared_path.rs (hunk 5 above)
Comment (sachindshinde) [reply apollo-federation/src/query_plan/incremental_planner/shared_path.rs]: (Noted similarly in [this comment](https://github.com/apollographql/router/pull/10101#discussion_r4001724735).) You could use `#[derive_where(Clone)]` to avoid the trait bound on `T`, via the [derive-where crate](https://crates.io/crates/derive-where).

**Review by dariuszkuc — APPROVED** [review]

File: apollo-federation/src/query_plan/incremental_planner/shared_path.rs (hunk 5 above)
Comment (tninesling) [reply apollo-federation/src/query_plan/incremental_planner/shared_path.rs]: Went with a manual `impl<T> Clone` (cloning only copies the Arc head, so no `T: Clone` bound) rather than pulling in derive-where; it's the same two derives' worth of code and avoids adding a dependency. Same rationale kept for the manual `Default`.

File: apollo-federation/src/query_plan/incremental_planner/shared_path.rs (hunk 6 above)
Comment (tninesling) [reply apollo-federation/src/query_plan/incremental_planner/shared_path.rs:142]: Kept the manual impl for the same reason as Clone above; a `derive_where` migration would only be worth it if we accumulate more of these.

File: apollo-federation/src/query_plan/incremental_planner/fetch_graph/selection_builder.rs (hunk 11 above)
Comment (tninesling) [reply apollo-federation/src/query_plan/incremental_planner/fetch_graph/selection_builder.rs]: Nothing downstream splits them; the keys are only compared whole, for merge-compatibility bucketing (that code lands later in the stack, in the debug-display PR). Switched the keys from joined strings to `Vec<String>` path segments there so nothing relies on a separator character at all.
